service/updates/downloader.go

```diff
@@ -0,0 +1,344 @@
+package updates
+
+import (
+	"archive/zip"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	semver "github.com/hashicorp/go-version"
+
+	"github.com/safing/portmaster/base/log"
+)
+
+type Downloader struct {
+	dir       string
+	indexFile string
+	indexURLs []string
+	bundle    *Bundle
+	version   *semver.Version
+
+	httpClient http.Client
+}
+
+func CreateDownloader(index UpdateIndex) Downloader {
+	return Downloader{
+		dir:       index.DownloadDirectory,
+		indexFile: index.IndexFile,
+		indexURLs: index.IndexURLs,
+	}
+}
+
+func (d *Downloader) downloadIndexFile(ctx context.Context) (err error) {
+	// Make sure dir exists
+	_ = os.MkdirAll(d.dir, defaultDirMode)
+	for _, url := range d.indexURLs {
+		err = d.downloadIndexFileFromURL(ctx, url)
+		if err != nil {
+			log.Warningf("updates: failed while downloading index file %s", err)
+			continue
+		}
+		// Downloading was successful.
+		err = nil
+		break
+	}
+
+	if err == nil {
+		err = d.parseBundle()
+	}
+
+	return
+}
+
+// Verify verifies if the downloaded files match the corresponding hash.
+func (d *Downloader) Verify() error {
+	err := d.parseBundle()
+	if err != nil {
+		return err
+	}
+
+	return d.bundle.Verify(d.dir)
+}
+
+func (d *Downloader) parseBundle() error {
+	indexFilepath := filepath.Join(d.dir, d.indexFile)
+	var err error
+	d.bundle, err = ParseBundle(indexFilepath)
+	if err != nil {
+		return err
+	}
+
+	d.version, err = semver.NewVersion(d.bundle.Version)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *Downloader) downloadIndexFileFromURL(ctx context.Context, url string) error {
+	// Request the index file
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, http.NoBody)
+	if err != nil {
+		return fmt.Errorf("failed to create GET request to %s: %w", url, err)
+	}
+	if UserAgent != "" {
+		req.Header.Set("User-Agent", UserAgent)
+	}
+
+	// Perform request
+	resp, err := d.httpClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed GET request to %s: %w", url, err)
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	// Check the status code
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return fmt.Errorf("received error from the server status code: %s", resp.Status)
+	}
+	// Create file
+	indexFilepath := filepath.Join(d.dir, d.indexFile)
+	file, err := os.Create(indexFilepath)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = file.Close() }()
+
+	// Write body of the response
+	_, err = io.Copy(file, resp.Body)
+	if err != nil {
+		return err
+	}
```

Comment:
Use `os.WriteFile` instead.

We will need to read the body in full anyway beforehand, to verify the signature, and to possibly parse the bundle and save it to the downloader as to not need to read the file again.